pre_commit_hooks/cfn_guard.py

```diff
@@ -0,0 +1,160 @@
+"""
+This module contains the logic for the cfn-guard pre-commit hook
+"""
+
+import json
+import os
+import platform
+import shutil
+import subprocess
+import sys
+import tarfile
+import tempfile
+from pathlib import Path
+from typing import Sequence, Union
+from urllib.request import Request, urlopen
+
+# pylint: disable=C0301
+LATEST_RELEASE_URL = (
+    "https://api.github.com/repos/aws-cloudformation/cloudformation-guard/releases/latest"
+)
+BIN_NAME = "cfn-guard"
+UNSUPPORTED_OS_MSG = "Unsupported operating system. Could not install cfn-guard."
+
+release_urls_dict = {
+    # pylint: disable=C0301
+    "darwin": "https://github.com/aws-cloudformation/cloudformation-guard/releases/download/TAG/cfn-guard-v3-macos-latest.tar.gz",
+    # pylint: disable=C0301
+    "linux": "https://github.com/aws-cloudformation/cloudformation-guard/releases/download/TAG/cfn-guard-v3-ubuntu-latest.tar.gz",
+    # pylint: disable=C0301
+    "windows": "https://github.com/aws-cloudformation/cloudformation-guard/releases/download/TAG/cfn-guard-v3-windows-latest.tar.gz",
+}
+supported_oses = ["linux", "darwin", "windows"]
+current_os = platform.system().lower()
+install_dir = os.path.join(os.path.expanduser("~"), ".cfn-guard-pre-commit")
+
+
+class CfnGuardPreCommitError(Exception):
+    """Custom exception class for specific error scenarios."""
+
+    def __init__(self, message, code=None):
+        """
+        Initialize the CfnGuardPreCommitError object.
+
+        Args:
+            message (str): The error message.
+            code (int, optional): An optional error code.
+        """
+        self.message = message
+        self.code = code
+        super().__init__(message)
+
+    def __str__(self):
+        """
+        Return a string representation of the CfnGuardPreCommitError object.
+        """
+        if self.code is not None:
+            return f"{self.message} (Code: {self.code})"
+        return self.message
+
+
+def request(url: str):
+    """Roll our own get request method to avoid extra dependencies"""
+
+    # Explicitly set the headers to avoid User-Agent "Python-urllib/x.y"
+    # https://docs.python.org/3/howto/urllib2.html#headers
+    return Request(url, headers={"User-Agent": "Mozilla/5.0"})
+
+
+def get_latest_tag() -> str:
+    """Get the latest release tag from Github"""
+
+    req = request(LATEST_RELEASE_URL)
+
+    with urlopen(req) as response:
+        data = response.read().decode("utf-8")
+        return json.loads(data)["tag_name"]
+
+
+def get_binary_name() -> str:
+    """Get an OS specific binary name"""
+
+    return BIN_NAME + (".exe" if current_os == "windows" else "")
+
+
+def install_cfn_guard():
+    """
+    Install the latest cfn-guard to the install_dir to avoid
+    global version conflicts with existing installations, rust,
+    and cargo.
+    """
+    latest_tag = get_latest_tag()
```

Comment:
I don't think this should be (or long term: stay) the latest tag, but the same version as the ref (which I realise is not easy).

Reasons:
    - If there is a major version release, rule files might stop working
    - currently no way to lock on a release version
    - That's how other pre-commit hooks work

One way to solve this might be reading the version from cargo.toml

Reply:
I like this! I tried to keep the tags separate since we use the non-prefixed versions of tags for the binary, the action- prefixed versions of tags for the github action, and the precommit- prefixed versions of tags for the precommit hook but tying the version to the one in the cargo.toml seems reasonable.